imap/sievedir.c

```diff
@@ -0,0 +1,426 @@
+}
+
+/* counts the number of scripts user has that are DIFFERENT from name */
+EXPORTED int sievedir_num_scripts(const char *sievedir, const char *name)
+{
+    struct count_rock crock = { 0, name };
+
+    sievedir_foreach(sievedir, &count_cb, &crock);
+
+    return crock.count;
+}
+
+
+EXPORTED struct buf *sievedir_get_script(const char *sievedir,
+                                         const char *script)
+{
+    struct buf buf = BUF_INITIALIZER;
+
+    buf_printf(&buf, "%s/%s", sievedir, script);
+
+    int fd = open(buf_cstring(&buf), 0);
+    if (fd < 0) return NULL;
+
+    buf_free(&buf);
+    buf_refresh_mmap(&buf, 1, fd, script, MAP_UNKNOWN_LEN, "sieve");
+
+    close(fd);
+
+    struct buf *ret = buf_new();
+
+    buf_move(ret, &buf);
+
+    return ret;
+}
+ 
+/* Everything but '/' and '\0' are valid. */
+EXPORTED int sievedir_valid_name(const struct buf *name)
+{
+    size_t lup, len = buf_len(name);
+    const char *ptr;
+
+    /* must be at least one character long */
+    if (len < 1) return 0;
+
+    ptr = buf_base(name);
+
+    for (lup = 0; lup < len; lup++) {
+        if ((ptr[lup] == '/') || (ptr[lup] == '\0')) return 0;
+    }
+
+    return (lup < 1013);
+}
+
+EXPORTED int sievedir_script_exists(const char *sievedir, const char *name)
+{
+    char path[PATH_MAX];
+    struct stat sbuf;
+
+    snprintf(path, sizeof(path), "%s/%s%s", sievedir, name, SCRIPT_SUFFIX);
+
+    return (stat(path, &sbuf) == 0);
+}
+
+EXPORTED const char *sievedir_get_active(const char *sievedir)
+{
+    static char target[PATH_MAX];
+    char link[PATH_MAX];
+    ssize_t tgt_len;
+
+    snprintf(link, sizeof(link), "%s/%s", sievedir, DEFAULTBC_NAME);
+
+    tgt_len = readlink(link, target, sizeof(target) - 1);
+
+    if (tgt_len > BYTECODE_SUFFIX_LEN) {
+        target[tgt_len - BYTECODE_SUFFIX_LEN] = '\0';
+        return target;
+    }
+    else if (tgt_len == -1 && errno != ENOENT) {
+        syslog(LOG_ERR, "IOERROR: readlink(%s): %m", link);
+    }
+
+    return NULL;
+}
+
+EXPORTED int sievedir_script_isactive(const char *sievedir, const char *name)
+{
+    if (!name) return 0;
+
+    return (strcmpnull(name, sievedir_get_active(sievedir)) == 0);
+}
+
+EXPORTED int sievedir_activate_script(const char *sievedir, const char *name)
+{
+    char target[PATH_MAX];
+    char active[PATH_MAX];
+    char tmp[PATH_MAX+4];  /* +4 for ".NEW" */
+
+    if (sievedir_script_isactive(sievedir, name)) {
+        /* already active - nothing to do here */
+        return SIEVEDIR_OK;
+    }
+
+    snprintf(target, sizeof(target), "%s%s", name, BYTECODE_SUFFIX);
+    snprintf(active, sizeof(active), "%s/%s", sievedir, DEFAULTBC_NAME);
+    snprintf(tmp, sizeof(tmp), "%s.NEW", active);
+
+    /* N.B symlink() does NOT verify target for anything but string validity,
+     * so activation of a nonexistent script will report success.
+     */
+    if (symlink(target, tmp) < 0) {
+        syslog(LOG_ERR, "IOERROR: unable to symlink %s as %s: %m", target, tmp);
```

Comment:
Can you do the IOERRORs with xsyslog please?  They're just gonna be converted anyway, might as well start out right.

Reply:
https://github.com/cyrusimap/cyrus-imapd/commit/534ad6f45f996076c98722207d62e9bf7f165fb8